File: studio/frontend/src/features/settings/api/linked-instances.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright 2026-present the Unsloth AI Inc. team. All rights reserved. See /studio/LICENSE.AGPL-3.0

import { authFetch } from "@/features/auth/api";

export interface LinkedInstance {
  id: string;
  name: string;
  base_url: string;
  created_at: string;
  updated_at: string;
}

export interface LinkedInstanceStatus {
  id: string;
  online: boolean;
  error: string | null;
  models: string[];
  loaded: string[];
  latency_ms: number | null;
}

async function detail(res: Response, fallback: string): Promise<Error> {
  try {
    const body = (await res.json()) as { detail?: unknown };
    if (typeof body.detail === "string") return new Error(body.detail);
  } catch {
    // not JSON
  }
  return new Error(fallback);
}

export async function fetchLinkedInstances(): Promise<LinkedInstance[]> {
  const res = await authFetch("/api/linked-instances");
  if (!res.ok) throw await detail(res, "Failed to load linked instances");
  return res.json();
}

export async function createLinkedInstance(input: {
  name: string;
  base_url: string;
  api_key: string;
}): Promise<LinkedInstance> {
  const res = await authFetch("/api/linked-instances", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
  if (!res.ok) throw await detail(res, "Failed to link instance");
  return res.json();
}

export async function deleteLinkedInstance(id: string): Promise<void> {
  const res = await authFetch(`/api/linked-instances/${id}`, {
    method: "DELETE",
  });
  if (!res.ok) throw await detail(res, "Failed to remove linked instance");
}

export async function testLinkedInstance(
  id: string,
): Promise<LinkedInstanceStatus> {
  const res = await authFetch(`/api/linked-instances/${id}/test`, {
    method: "POST",
  });
  if (!res.ok) throw await detail(res, "Failed to reach linked instance");
  return res.json();
}

export async function fetchLinkedInstancesStatus(): Promise<
  LinkedInstanceStatus[]
> {
  const res = await authFetch("/api/linked-instances/status");
  if (!res.ok) throw await detail(res, "Failed to load linked instances");
  return res.json();
}

export async function updateLinkedInstance(
  id: string,
  input: { name?: string; base_url?: string; api_key?: string },
): Promise<LinkedInstance> {
  const res = await authFetch(`/api/linked-instances/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
  if (!res.ok) throw await detail(res, "Failed to update linked instance");
  return res.json();
}

export interface LinkedInstanceGpu {
  name: string;
  vram_total_gb: number | null;
  vram_used_gb: number | null;
  utilization_pct: number | null;
}

/** What a linked instance reports about itself; older releases leave fields null. */
export interface LinkedInstanceInfo {
  id: string;
  online: boolean;
  error: string | null;
  version: string | null;
  install_source: string | null;
  update_available: boolean;
  latest_version: string | null;
  platform: string | null;
  python_version: string | null;
  device_backend: string | null;
  torch: string | null;
  transformers: string | null;
  cuda: string | null;
  rocm: string | null;
  llama_cpp: string | null;
  gpus: LinkedInstanceGpu[];
  cpu_count: number | null;
  memory_total_gb: number | null;
  memory_available_gb: number | null;
  disk_total_gb: number | null;
  disk_free_gb: number | null;
  uptime_seconds: number | null;
}

export async function fetchLinkedInstancesInfo(): Promise<
  LinkedInstanceInfo[]
> {
  const res = await authFetch("/api/linked-instances/info");
  if (!res.ok) throw await detail(res, "Failed to load linked instances");
  return res.json();
}

File: studio/backend/models/linked_instances.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright 2026-present the Unsloth AI Inc. team. All rights reserved. See /studio/LICENSE.AGPL-3.0

"""Pydantic schemas for the linked Unsloth Studio instances API."""

from typing import Optional

from pydantic import BaseModel, Field


class LinkedInstanceCreate(BaseModel):
    name: str = Field(..., max_length = 32, description = "Short name; models appear as @<name>/<model>")
    base_url: str = Field(..., max_length = 2048, description = "The remote's URL, e.g. its trycloudflare.com address")
    api_key: str = Field(..., min_length = 1, max_length = 512, description = "An API key created on the remote")


class LinkedInstanceUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length = 32)
    base_url: Optional[str] = Field(None, max_length = 2048)
    api_key: Optional[str] = Field(None, max_length = 512)


class LinkedInstance(BaseModel):
    id: str
    name: str
    base_url: str
    created_at: str
    updated_at: str


class LinkedInstanceStatus(BaseModel):
    id: str
    online: bool
    error: Optional[str] = None
    models: list[str] = Field(default_factory = list, description = "Model ids as this server exposes them")
    loaded: list[str] = Field(default_factory = list, description = "The subset currently loaded on the remote")
    latency_ms: Optional[int] = None


class LinkedInstanceGpu(BaseModel):
    name: str
    vram_total_gb: Optional[float] = None
    vram_used_gb: Optional[float] = None
    utilization_pct: Optional[float] = None


class LinkedInstanceInfo(BaseModel):
    """What the remote reports about itself; any field can be missing on older releases."""

    id: str
    online: bool
    error: Optional[str] = None
    version: Optional[str] = None
    install_source: Optional[str] = None
    update_available: bool = False
    latest_version: Optional[str] = None
    platform: Optional[str] = None
    python_version: Optional[str] = None
    device_backend: Optional[str] = None
    torch: Optional[str] = None
    transformers: Optional[str] = None
    cuda: Optional[str] = None
    rocm: Optional[str] = None
    llama_cpp: Optional[str] = None
    gpus: list[LinkedInstanceGpu] = Field(default_factory = list)
    cpu_count: Optional[float] = None
    memory_total_gb: Optional[float] = None
    memory_available_gb: Optional[float] = None
    disk_total_gb: Optional[float] = None
    disk_free_gb: Optional[float] = None
    uptime_seconds: Optional[float] = None
